server/src/main/java/io/unitycatalog/server/service/AuthService.java

```diff
@@ -0,0 +1,158 @@
+    String CLIENT_SECRET = "client_secret";
+    String SUBJECT_TOKEN = "subject_token";
+    String SUBJECT_TOKEN_TYPE = "subject_token_type";
+    String ACTOR_TOKEN = "actor_token";
+    String ACTOR_TOKEN_TYPE = "actor_token_type";
+    String REQUESTED_TOKEN_TYPE = "requested_token_type";
+  }
+
+  interface GrantTypes {
+    String TOKEN_EXCHANGE = "urn:ietf:params:oauth:grant-type:token-exchange";
+  }
+
+  interface TokenTypes {
+    String ACCESS = "urn:ietf:params:oauth:token-type:access_token";
+    String ID = "urn:ietf:params:oauth:token-type:id_token";
+    String JWT = "urn:ietf:params:oauth:token-type:jwt";
+  }
+
+  interface AuthTypes {
+    String BEARER = "Bearer";
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AuthService.class);
+
+  private final SecurityContext securityContext;
+  private final JwksOperations jwksOperations;
+
+  public AuthService(SecurityContext securityContext) {
+    this.securityContext = securityContext;
+    this.jwksOperations = new JwksOperations();
+  }
+
+  /**
+   * OAuth token exchange.
+   *
+   * <p>Performs an OAuth token exchange for an access-token.
+   *
+   * <p>TODO: This could probably be integrated into the OpenAPI spec.
+   *
+   * @param request The token exchange parameters
+   * @return The token exchange response
+   */
+  @Post("/tokens")
+  public HttpResponse grantToken(OAuthTokenExchangeRequest request) {
+    LOGGER.debug("Got token: {}", request);
+
+    if (request.getGrantType() != null
+        && !TokenTypes.ACCESS.equals(request.getRequestedTokenType())) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT,
+          "Unsupported requested token type: " + request.getRequestedTokenType());
+    }
+
+    if (request.getSubjectTokenType() == null) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT, "Subject token type is required but was not specified");
+    }
+
+    if (request.getActorTokenType() != null) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT, "Actor tokens not currently supported");
+    }
+
+    DecodedJWT decodedJWT = JWT.decode(request.getSubjectToken());
+    String issuer = decodedJWT.getClaim("iss").asString();
+    String keyId = decodedJWT.getHeaderClaim("kid").asString();
+
+    LOGGER.debug("Validating token for issuer: {}", issuer);
+
+    JWTVerifier jwtVerifier = jwksOperations.verifierForIssuerAndKey(issuer, keyId);
+    decodedJWT = jwtVerifier.verify(decodedJWT);
+
+    String accessToken = securityContext.createAccessToken(decodedJWT);
+
+    OAuthTokenExchangeResponse response =
+        OAuthTokenExchangeResponse.builder()
+            .accessToken(accessToken)
+            .issuedTokenType(TokenTypes.ACCESS)
+            .tokenType(AuthTypes.BEARER)
+            .build();
+
+    return HttpResponse.ofJson(response);
+  }
+
+  // TODO: This should be probably integrated into the OpenAPI spec.
+  @ToString
```

Comment:
Nit: Can you better document the API -- what's being exchanged for what etc.,? 

IIUC, the `subjectToken`, which could be an OAuth access token from an external IdP (e.g., Google) is being exchanged for an access token by the local issuer for the same email subject -- is that correct?

Reply:
I added some javadoc to the `grantToken()` method. Yes, you are correct, the `subjectToken` would be an id or access token from an IdP.